### ISSUE TITLE
Some hints are being obscured by brackets

### ISSUE THREAD

**Issue by yairchu** [ndmitchell/hlint#1183]
Since using the new "Redundant bracket due to operator fixities" option, I found that sometimes after removing redundant brackets, other rules which were previously obscured by them are triggered.

Here's a real example: https://github.com/Peaker/git-mediate/commit/78f09c611794c920dcdb9cf71779df7a48a4e6a4

Should `hlint` internally remove redundant brackets so that it could better check existing rules?

**Comment by ndmitchell** [comment]
HLint should be able to see through most brackets - ignoring even brackets which are required. Would be good to have a minimal example of whats going on there, what rule it triggers and what brackets mean it doesn't.

**Comment by yairchu** [comment]
So perhaps the issue is that rules should be written with brackets and some aren't? For example I see that

```yaml
- hint: {lhs: (f `on` g) `on` h, rhs: f `on` (g . h), name: Fuse on/on}
```

has (redundant iiuc) brackets in the rule. Will try to make a minimal example.

**Comment by ndmitchell** [comment]
Most rules don't have brackets in. In general, if there is a bracket in the expression the user wrote, but not in the LHS of the rule we match. If there is a bracket in the LHS but not in the user expression, we don't. But it's a somewhat complex area, so there are some special cases (which I don't understand fully).

**Comment by yairchu** [comment]
Here's a small example:

```Haskell
x00 = a >= 'a' && a <= 'z'
x01 = a >= 'a' && (a <= 'z')
x10 = (a >= 'a') && a <= 'z'
x11 = (a >= 'a') && (a <= 'z')
```

The hint

```yaml
- warn: {lhs: a >= 'a' && a <= 'z', rhs: isAsciiLower a}
```

triggers only for `x00`.

But if we add redundant brackets in the hint:

```yaml
- warn: {lhs: (a >= 'a') && (a <= 'z'), rhs: isAsciiLower a}
```

then it triggers for all of the variants.

**Comment by ndmitchell** [comment]
Thanks for providing the minimal example, much easier to track down!

We were removing brackets on the rule side, but not the user side. I now remove them on both sides, and that makes the above work without causing any regressions. So it looks like the next HLint will match rules in more places.
